src/buildercore/core.py

```diff
@@ -243,10 +243,12 @@ def stack_pem(stackname, die_if_exists=False, die_if_doesnt_exist=False):
     return expected_key
 
 def _ec2_connection_params(stackname, username, **kwargs):
-    "returns a dictionary of settings to be used with Fabric's api.settings context manager"
+    "returns a dictionary of settings to be used with `command.settings` context manager"
     # http://docs.fabfile.org/en/1.14/usage/env.html
     params = {'user': username}
     pem = stack_pem(stackname)
+    # todo: handle other non-rsa private keys
```

Comment:
ignores `pem`? What's the preference between the two?

Reply:
do you mean `id_rsa.pem` and `id_rsa.pub` ? `ssh-keygen` doesn't output a `.pem` file.

I think this change to `_ec2_connection_params` can be reverted and the `key_filename` pushed back into `command.py` and the call to `threadbare.state.set_defaults`. `set_defaults` was implemented only recently.

Reply:
> ignores pem? What's the preference between the two?

oh, I think I see what you mean now. It looks like I was setting a default. Fabric has *loads* of defaults, the path to a private key among them. Threadbare starts with no defaults. There is a case here where the bootstrap user's `pem` key isn't added because the expected file doesn't exist. Fabric would default to the user's key, Threadbare would fail.

I've moved this particular default into `command.py`. I'll also expand it to check several other common keys. I've seen implementations in both Fabric and PSSH for that